src/rust/src/contract.rs

```diff
@@ -2,31 +2,154 @@
 use std::io::Cursor;
 
-use cxx::{CxxString, CxxVector};
+use cxx::CxxString;
+use im_rc::OrdMap;
 use std::error::Error;
 use stellar_contract_env_host::{
-    xdr::{ReadXdr, ScVec, WriteXdr},
-    Host, VM,
+    storage::{self, AccessType, Key, Storage},
+    xdr,
+    xdr::{
+        ContractDataEntry, Hash, LedgerEntry, LedgerEntryData, LedgerKey, ReadXdr, ScObject,
+        ScStatic, ScVal, ScVec, WriteXdr,
+    },
+    Host, HostError, Vm,
 };
 
-/// Deserialize an SCVec XDR object of SCVal arguments from the C++ side of the
-/// bridge, instantiate a Host and VM with the provided WASM, invoke the
-/// requested function in the WASM, and serialize an SCVal back into a return
-/// value.
+/// Helper for [`build_storage_footprint_from_xdr`] that inserts a copy of some
+/// [`AccessType`] `ty` into a [`storage::Footprint`] for every [`LedgerKey`] in
+/// `keys`.
+fn populate_access_map_with_contract_data_keys(
+    access: &mut OrdMap<Key, AccessType>,
+    keys: Vec<LedgerKey>,
+    ty: AccessType,
+) -> Result<(), HostError> {
+    for lk in keys {
+        match lk {
+            LedgerKey::ContractData(xdr::LedgerKeyContractData { contract_id, key }) => {
+                let sk = Key { contract_id, key };
+                access.insert(sk, ty.clone());
+            }
+            _ => return Err(HostError::General("unexpected ledger key type").into()),
+        }
+    }
+    Ok(())
+}
+
+/// Deserializes an [`xdr::Footprint`] from the provided [`XDRBuf`], converts
+/// its entries to an [`OrdMap`], and returns a [`storage::Footprint`]
+/// containing that map.
+fn build_storage_footprint_from_xdr(footprint: &XDRBuf) -> Result<storage::Footprint, HostError> {
+    let xdr::Footprint {
+        read_only,
+        read_write,
+    } = xdr::Footprint::read_xdr(&mut Cursor::new(footprint.data.as_slice()))?;
+    let mut access = OrdMap::new();
+
+    populate_access_map_with_contract_data_keys(
+        &mut access,
+        read_only.to_vec(),
+        AccessType::ReadOnly,
+    )?;
+    populate_access_map_with_contract_data_keys(
+        &mut access,
+        read_write.to_vec(),
+        AccessType::ReadWrite,
+    )?;
+    Ok(storage::Footprint(access))
+}
+
+/// Deserializes a sequence of [`xdr::LedgerEntry`] structures from a vector of
+/// [`XDRBuf`] buffers, inserts them into an [`OrdMap`] with entries
+/// additionally keyed by the provided contract ID, checks that the entries
+/// match the provided [`storage::Footprint`], and returns the constructed map.
+fn build_storage_map_from_xdr_ledger_entries(
+    footprint: &storage::Footprint,
+    ledger_entries: &Vec<XDRBuf>,
+) -> Result<OrdMap<Key, Option<ScVal>>, HostError> {
+    let mut map = OrdMap::new();
+    for buf in ledger_entries {
+        let le = LedgerEntry::read_xdr(&mut Cursor::new(buf.data.as_slice()))?;
+        match le.data {
+            LedgerEntryData::ContractData(ContractDataEntry {
+                key,
+                val,
+                contract_id,
```

Comment:
If I understood correctly, here the ledger entry is destructured into three members of the struct. Shouldn't the order be `contract_id`, `key`, `val`?

Reply:
They're destructured by field name, not position.